control/control.rnc

```diff
@@ -566,9 +566,14 @@ system_role = element system_role {
     ## texts section: ${id} and ${id}_description.
     element id { text } &
 
-    ## option to not preselect any role, by default first one is preselected
+    ## option to not preselect any role, by default first one by order is preselected
     element no_default { BOOLEAN }? &
 
+    ## Option for order of role. Roles are sorted by order, the smaller first. It can be negative.
+    ## If not specified default 10 000 is used. If two roles have same order, then the first come
+    ## rule is used, so it is stable sorting.
```

Comment:
"stable" from our code POV, but in reality it might be different. It might depend in which order the add-ons are added, how the packages are sorted internally by libzypp, etc...

I'd say the order is undefined for the roles with the same ordering number. If somebody really wants to have a defined order then unique numbers should be used. The range is quite wide to avoid collisions.

Reply:
well, I write there first come. I want to mention it there, so e.g. openSUSE which does not need to use order, can simply keep it as it is. Of course if original order is not well defined, then stable sorting does not help.

Reply:
OK, let's hope the "first come" term is understandable...

Reply:
I would not rely on any order if none is specified. Maybe it would be better to use some other field if the numeric order is the same; say, sort first by order and then by the (translated, if possible) role name. Not sure  if this is academic because we very likely don't have so many that a product/project manager in charge of such a product wouldn't define his custom order explicitly in every case.

Also, again: Is it possible for those roles to come from two different places like in the example above, some from the SLE base product and some from an add-on product? I guess with the advent of LeanOS + add-on modules this might easily happen. In that case, what exactly is the "first appearance"? LeanOS and then the modules - in what order?

How about sorting by role name alphabetically as the secondary sort criterium?

Reply:
@shundhammer well, that is something I would like to avoid, as it means we have to change all control files that use roles. E.g. opensuse does have it only in control file and order is order in control file. Maybe I can write that if roles is added from add-ons then order is not defined and always use order key? But in that case we have again problem with nil, because it will mean that if original project does not use order and add-on use it, then add-on ones comes always as the first ones.

Reply:
I think that it might even be preferable for the product makers to have their own roles on top; they have specialized roles which is very likely one key reason for customers to buy that product in the first place.

But yes, clearly writing that "always use orders to avoid unpleasant surprises" is very likely the most pragmatic approach.